tv_apps/tv_shared/js/animations.js

```diff
@@ -72,6 +72,115 @@
           return isPlaying;
         }
       };
+    },
+
+    _getChild: function(container, childRelation, index) {
+      var targetElem = container.children[index];
+      childRelation = childRelation.child;
+      while(childRelation.child) {
+        targetElem = targetElem.children[0];
+        childRelation = childRelation.child;
+      }
+      return targetElem;
+    },
+
+    _findChildInViewport: function(container, childRelation) {
+      // find the elements in viewport
+      var child;
+      var childRect;
+      var childrenInViewport = [];
+      var i;
+      for(i = 0; i < container.children.length; i++) {
+        child = this._getChild(container, childRelation, i);
+        childRect = child.getBoundingClientRect();
+        if (childRect.right >= 0 &&
+          childRect.left <= window.innerWidth) {
+          childrenInViewport.push({
+            left: childRect.left,
+            child: child
+          });
+        } else {
+          child.classList.remove('animation-bubble-init');
+        }
+      }
+
+      // sort child indexes. Card position may not be the same as index
+      // after a new child is pinned to home
+      childrenInViewport.sort(function(childObj1, childObj2) {
+        if (childObj1.left > childObj2.left) {
+          return 1;
+        }
+        if (childObj1.left < childObj2.left) {
+          return -1;
+        }
+        return 0;
+      });
+
+      return childrenInViewport;
```

Comment:
The actual order of cards are preserved in scrollable objects. (specifically, index of scrollable._nodes[] corresponds to actual positions.)
Would it be easier to use of `scrollable.getNextItem`, `scrollable.getPrevItem`, and `scrollable.getNode`?